TGR-102: Updating ruby to 3.3, docs, and minor aws gems implementation

EdwinGuzman opened the pull request:
Resolves [TGR-102](https://newyorkpubliclibrary.atlassian.net/browse/TGR-102)

While I got the repo to work locally following the instructions in the readme using ruby 3.3, I'm not entirely sure how to go about testing this in QA or how other services are related to this.

- I'm not familiar with the ruby and gem ecosystem so please let me know if I need to make any updates in the gem files.
- Tests run and pass.
- Localstack had minor updates since this was last updated. I had to create a new encrypted `SQS_QUEUE_URL` variable to use port "4566" instead of "4567".
- Travis is replaced by Github Actions for running tests and the deployment process.

[TGR-102]: https://newyorkpubliclibrary.atlassian.net/browse/TGR-102?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

EdwinGuzman committed:
Updating ruby to 3.3, docs, and minor aws gems implementation

croyfish commented on Gemfile:
```diff
@@ -1,7 +1,8 @@
+gem 'aws-sdk-sqs', '~> 1.81'
+gem 'aws-sdk-core', '~> 3.203'
+gem 'aws-sdk-kms', '~> 1.89'
 gem 'nypl_log_formatter', '~> 0.1.2'
```
These upgrades seem reasonable. What I have been doing (for example, with [mms](https://github.com/NYPL/mms/blob/qa/Gemfile)) is leaving required versions off entirely, removing Gemfile.lock, and then regenerating it with `bundle install` to get the latest versions. 

By not adding version constraints, we should be able to keep dependencies up to date without having to think about it too much, and if we find a case where a version needs to be locked, then we can do it at that point.

I was wondering a bit about nypl_log_formatter. It looks like we could actually upgrade it to [version 0.1.3](https://rubygems.org/gems/nypl_log_formatter/versions), but that dependency itself hasn't been updated in many years (edit: I see in the Gemfile.lock that we are pulling in 1.3). nypl_platform_api_client is likewise > 4 years old, but we're already on the latest version of that.

EdwinGuzman replied:
I removed the versions from the aws-related gems but left it for the nypl-related gems.

croyfish reviewed (approved):
Went through the local steps as well, and it seems to be working with the proposed changes:

```
2024-09-11T15:47:23.021  INFO --- [et.reactor-0] localstack.request.aws     : AWS sqs.ReceiveMessage => 200
2024-09-11T15:49:01.793  INFO --- [et.reactor-0] localstack.request.aws     : AWS sqs.ReceiveMessage => 200
```

Similarly, `bundle exec rspec` shows no failures. You're right that other expected behavior is not well documented here, and not having everything running in docker containers is a little rough, but on the whole, I think we should be fine with these upgrades.

EdwinGuzman committed:
removing strict version in gemfile

EdwinGuzman committed:
Adding github actions and terraform process, and removing travis.

EdwinGuzman committed:
typo

EdwinGuzman committed:
installing gems

EdwinGuzman committed:
installing gems

EdwinGuzman commented:
@croyfish thanks for the review. I just removed the "WIP" and replaced travis with github actions and terraform. It'll have to be merged into `qa` in order to test that the deployment works but this follows the same pattern as other LSP lambdas.

EdwinGuzman committed:
wrong env name in tf file

charmingduchess commented on provisioning/base/resources.tf:
```diff
@@ -0,0 +1,50 @@
+  type        = "zip"
+  output_path = "${path.module}/dist.zip"
+  source_dir  = "../../"
+  excludes    = [".git", ".terraform", "provisioning"]
```
probably can include spec files in this array?

charmingduchess reviewed (commented):
Looks good to me! I also do not have much insight about the ruby upgrade details.

danamansana commented on .github/workflows/test-and-deploy.yml:
```diff
@@ -0,0 +1,59 @@
+      id-token: write
+      contents: read
+    needs: test
+    if: github.ref == 'refs/head/qa'
```
I've been using `tgr-qa` as a temporary measure but this is fine if you are comfortable using the regular `qa` branch

EdwinGuzman replied:
I think this should be ok to merge directly and then test or do you prefer testing the PR and then merging it? We'd have to update the lambda to use the ruby 3.3 runtime, but if it fails, I don't think we can revert to a lower ruby runtime.

danamansana commented on provisioning/base/resources.tf:
```diff
@@ -0,0 +1,50 @@
+  region = "us-east-1"
+}
+
+variable "environment" {
```
How is this app handling environment variables? Other lambdas that use terraform use something like this: 

`# Load ENV vars from ./config/{environment}.env
  environment {
    variables = { for tuple in regexall("(.*?)=(.*)", file("../../config/${var.environment}.env")) : tuple[0] => tuple[1] }
  }`

nonword replied:
Looks like previously app config was only stored in `.travis.yml`, which is removed in this PR. The env vars should prob be moved to `./config/*.env` files and loaded in tf as @danamansana suggests above.

danamansana reviewed (approved):
A couple of minor comments/questions but looks good

nonword commented on README.md:
````diff
@@ -70,14 +70,15 @@ aws_secret_access_key =
-```
-aws sqs receive-message --region us-east-1 --queue-url http://localhost:4576/queue/proxy-request-service --endpoint http://localhost:4576 --profile local --attribute-names All --message-attribute-names All --max-number-of-messages 10
+```sh
+aws sqs receive-message --region us-east-1 --queue-url http://localhost:4566/000000000000/proxy-request-service --endpoint http://localhost:4566 --profile local --attribute-names All --message-attribute-names All --max-number-of-messages 10
````
Is that `000000000000` intentional?

EdwinGuzman replied:
Yes, that's the new url structure from localstack

EdwinGuzman committed:
Adding env var in config files for terraform

EdwinGuzman commented:
Updated the env vars in terraform

nonword reviewed (approved):
Looks right to me. I can assist testing in QA if you like.

croyfish commented on lib/kms_client.rb:
```diff
@@ -3,7 +3,11 @@
+      region: 'us-east-1',
+      access_key_id:  ENV['AWS_ACCESS_KEY_ID'],
+      secret_access_key: ENV['AWS_SECRET_ACCESS_KEY']
+    )
```
Was this change a deliberate choice, or did you find that it was required?

EdwinGuzman replied:
Typically, the aws creds from the `~/.aws/credentials` file are used implicitly when the app runs. I found that in this docker setup, that wasn't true and I had to pass them in explicitly.
